### ISSUE TITLE
DHCP feature request: allow reservations outside of scope

### ISSUE THREAD

**Issue by pricerc** [TechnitiumSoftware/DnsServer#831]
Currently, when adding a reservation, it must be part of the address range for the scope.

This seems like an unnecessary limitation. 

I understand that it needs to be in the same subnet, but OpenWRT (for example) allows the reservation to be any IP.

In my case, I now have Technitium DNS/DHCP servers running on two Raspberry Pi 4's and would like to have non-overlapping DHCP scopes but with shared reservations for some systems (e.g. printer) on my network.

While I'm here. For the user-interface, I think the reservation entries should "lock" the subnet part of the IP address - although I realise that could get tricky if someone is using an 'odd' subnet boundary.

**Comment by pricerc** [comment]
> Can you elaborate this a bit? Do you mean the subnet field in the scope at the top should be locked?

No, I mean the IP address for a reservation should not allow you to even type an invalid address.

When adding reservations, the IP address is a free text field. 

So, even though my subnet is, say 10.10.0.0/16, there is nothing to stop me putting in an address of 10.100.2.3.

Ideally, the "10.10." part (the subnet) should be "locked" when I'm adding a new reservations, but I can't think of a good UX for that kind of behaviour. Especially if you want to support classless addressing, where someone could have a 10.10.0.0/20 subnet...

So perhaps it could just be validated and highlighted in javascript at the time of entry, rather than waiting until the admin clicks the 'save' button.

**Comment by pricerc** [comment]
> Will explore this thing once and see how it can be done.

I should note that for my scenario, 'clustering' would be a more convenient scenario - I have two systems that I'd ideally like to have sharing a database.

**Comment by ShreyasZare** [comment]
> So perhaps it could just be validated and highlighted in javascript at the time of entry, rather than waiting until the admin clicks the 'save' button.

This will need to implement the network address validation code in JS which is good to have and can be done later once pending features are implemented.

**Comment by ShreyasZare** [comment]
> > Will explore this thing once and see how it can be done.
> 
> I should note that for my scenario, 'clustering' would be a more convenient scenario - I have two systems that I'd ideally like to have sharing a database.

Clustering is planned feature. First the clustering support that is planned to be implemented will cover DNS parts only. Once that is settled that DHCPv6 and clustering support for it will be implemented.

**Comment by IconicLight** [comment]
I would like to add that I'm looking for this feature, in mostly the way described.

For instance I would like to assign a client an IP of 192.168.100.10, however my DHCP scope range is set to 192.168.100.200 to 192.168.100.250.

I'm ok with all octets being enforced except for the last one. I would like to preferably be able to set a DHCP static lease to a client outside of the 192.168.100-200-250 range which should be used by DHCP clients.

I've moved my DHCP from OPNsense to Technitium. OPNsense allows you to set a static lease for any IP on the last octet, would love if Technitium could do this feature as well.

**Comment by brokoler** [comment]
Indeed a useful feature, until it gets added there is a workaround.

1. For the Scope range, enter your full subnet:

![grafik](https://github.com/TechnitiumSoftware/DnsServer/assets/57031979/e5cfb1fe-ef98-42e8-981b-346e928c15b3)

2. Exclude the ranges you would not like to serve automatically via DHCP:

![grafik](https://github.com/TechnitiumSoftware/DnsServer/assets/57031979/e6f34bb8-5f90-4737-be5d-f59a3d9896ff)

3. Add the reserved leases outside the usual DHCP scope, which is part of your defined exclusion range:

![grafik](https://github.com/TechnitiumSoftware/DnsServer/assets/57031979/439b3223-4459-4b16-b858-e8ac27e9dbf5)

Works without any problems for me.
Maybe this is useful for someone else.

**Comment by christianstrauch** [comment]
Here's how I solved what I believe is similar to the OP's problem:

My requirements:
- There shall be enough IPv4 addresses in my local network - I chose a 255.255.252.0 subnet.
- Three different types of devices shall be in three different IP Class C subnets, and in three different local subdomains in DNS (.example.local, iot.example.local, dyn.example.local)
- One of these three different types of devices shall be temporary DHCP leases (e.g., client computers, phones, TVs, ...,) they shall be put into the dyn.example.local domain; static leases for IOT devices (e.g., smart plugs, blinds, ...) shall be put into the iot.example.local domain; well known servers shall be in the example.local domain.

My solution:
- Created three DHCP scopes with "Starting Address" and "Ending Address" defined in the respectively required class C subnets and the 255.255.252.0 subnet mask
- Put the static leases proactively into the iot and servers scope using the devices' MAC addresses
- Checked "Allow Only Reserved Lease Allocations" for the iot and servers scope
- Set the "Offer delay time" to 0 for servers, 50ms IOT devices, and 100ms for the dynamic scope (although I should now be able to set 0 for IOT as well; I had to do this staggering of servers and IOT for migration purposes)
- Set the "Domain name" for the scopes to example.local, iot.example.local, dyn.example.local, respectively

This way, DHCP offers the reserved server leases first, then the reserved IOT leases, then dynamic leases as a fallback.

**Comment by e-p-s** [comment]
> I would like to add that I'm looking for this feature, in mostly the way described.
> 
> For instance I would like to assign a client an IP of 192.168.100.10, however my DHCP scope range is set to 192.168.100.200 to 192.168.100.250.
> 
> I'm ok with all octets being enforced except for the last one. I would like to preferably be able to set a DHCP static lease to a client outside of the 192.168.100.200-250 range which should be used by DHCP clients.
> 
> I've moved my DHCP from OPNsense to Technitium. OPNsense allows you to set a static lease for any IP on the last octet, would love if Technitium could do this feature as well.

Me the same. OpnSense -> Technitium
this feature is badly missed

**Comment by JPLab90** [comment]
Are there any news? The post is nearly one year old. I am also currently working on switching from opnsense to technitium and like to see this feature :-)

**Comment by ShreyasZare** [comment]
> Are there any news? The post is nearly one year old. I am also currently working on switching from opnsense to technitium and like to see this feature :-)

@JPLab90 Thanks for asking. This will take some time since there are a few DNS features planned before work on the major upgrade for DHCP server is taken for implementation.

**Comment by CoryCharlton** [comment]
Some concrete findings from v15.3 (`bfc99287`):

The API/UI asymmetry is real: the in-range check is only in the `Scope.ReservedLeases` setter (driven by the console form via `dhcp/scopes/set`); `addReservedLease` → `Scope.TryAddReservedLease` does no range check, so the API creates out-of-range reservations the UI rejects.

But a UI-only relaxation would be a footgun — which I think is your "scope selection during renewal" point. Reserved leases key on MAC, so DISCOVER (`ciaddr=0`) hits the broadcast branch of `FindScope` and the out-of-range reservation is assigned fine. On renewal the client unicasts a REQUEST with `ciaddr` = its out-of-range IP → the `else` branch of `FindScope` selects the scope via `IsAddressInRange(ciaddr)` → no scope matches → `FindScope` returns null → the server drops the request (no NAK). So out-of-range reservations work only until first renewal, then the client loses the lease and re-DISCOVERs.

A correct fix looks like two parts: relax the `ReservedLeases` setter to validate subnet membership instead of pool range, and make the renewal branch of `FindScope` locate the scope by client-id / network membership (as the broadcast and relay branches already do) rather than `IsAddressInRange(ciaddr)`. 

@ShreyasZare, I'm happy to put up a PR — but given the planned DHCP rework, is that close enough that a patch against current code would just be thrown away?

**Comment by CoryCharlton** [comment]
Confirmed on a production fleet (v15.3), all in one scope (pool `10.0.1.100–.250`, 1-day lease):

- Reserved leases **inside** the pool (`.200`, `.201`) and dynamic clients: `leaseExpires` advances normally — renewing fine.
- Reserved leases **outside** the pool (`.20`–`.28`): `leaseExpires` stays frozen at the last server restart and never advances past T1; they only get a fresh lease by re-DISCOVERing at expiry (so `leaseObtained` resets each cycle).

Same lease type — they differ only by in/out of pool — so it reproduces exactly at the `FindScope` renewal branch (`IsAddressInRange(ciaddr)` finds no scope → the unicast REQUEST is dropped). Offer to PR the two-part fix still stands if the DHCP rework isn't close.

**Comment by CoryCharlton** [comment]
The two-part fix turned out to be small and self-contained (2 files, ~3 hunks), so rather than wait I went ahead and opened #2032.

It validates reserved addresses by subnet membership instead of pool range (in both the `ReservedLeases` setter and the `addReservedLease` API path), and makes the no-relay unicast RENEW/REBIND branch of `FindScope` match the scope by the client's reserved lease (MAC) before the pool-range check — mirroring the broadcast and relay branches that already work. Verified end-to-end with an isolated veth/netns DHCP exchange: out-of-pool reservation offered/ACKed on DISCOVER and, crucially, ACKed on unicast renewal (previously dropped at T1); wrong-subnet addresses still rejected on both entry points.

Understood if it needs to wait on the planned DHCP rework — happy to iterate or rebase whenever you get a chance to look.

**Comment by ShreyasZare** [comment]
@CoryCharlton Thanks for the analysis. The add reserved lease API is indeed missing the check but it is not being used with the GUI.

The DHCP server is planned to be completely redesigned so any changes to how it works is not really something I am looking forward to. The work on this redesign may start in 2-3 months once a couple of pending DNS features are implemented.

**Comment by CoryCharlton** [comment]
@ShreyasZare, ok well the fix was small so https://github.com/TechnitiumSoftware/DnsServer/pull/2032 is still available in the interim. Let me know if you'd rather I close it and wait for the redesign.
